### ISSUE TITLE
Error on unsafe comparison / spaceship operator

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#6296]
IMO, all of these operators: `<`, `>`, `<=`, `>=`, `<=>` should only be used on a handful of types.

Some comparisons just don't make much sense: what should `1 < null` really return? What about `1 <=> ['hello']`? Even though such comparisons are allowed in PHP, **I think they're most of the time a sign of a potential bug in an application**, and should be reported by Psalm by default.

Same thing with objects: the [docs](https://www.php.net/manual/en/language.oop5.object-comparison.php) don't seem to explain how objects are compared with these operators, and even if they did, I believe that in the absence of a proper interface to let objects compare themselves, relying on object comparison is *very* dangerous, apart maybe from internal objects that seem to be designed for this purpose, such as `DateTime`/`DateTimeImmutable`.

Same thing with arrays, there aren't many valid cases of array comparison, and even if a handful of those exists somewhere in an obscure library, that could deserve a `@psalm-suppress` IMO.

**Would it be possible to implement a new error in Psalm, when the comparison is considered dangerous?**

Some ideas of what would be considered OK, and what wouldn't be:

| Type A | Type B | Result of `<`, `>`, `<=`, `>=`, `<=>`                             |
|--------|--------|--------------------------------------|
| int    | int    | OK                                   |
| string | string | OK                                   |
| int    | string | OK (not ideal, but quite well defined in PHP now)                                  |
| object | object | ERROR, unless both `DateTimeInterface`? |
| any other combination |   | ERROR                                 |

Some examples: https://psalm.dev/r/ca87ef87b9

Thoughts?

**Comment by weirdan** [comment]
Love the idea, but would really like to see results with a selection of projects using Psalm, before considering merging anything like this.

**Comment by BenMorel** [comment]
@orklah Good point. Let's see how this discussion goes: https://externals.io/message/115648
